### PR TITLE
PPS return fixes

### PR DESCRIPTION
**Pull request by torotil**
Paypal seems to have stopped sending proper IPN packets along-side the return request (see [#2919709](https://www.drupal.org/project/paypal_payment/issues/2918709)). Need to compensate …

### CODE REVIEW

File: paypal_payment_pps/paypal_payment_pps.module [hunk 1]
```diff
@@ -247,26 +251,28 @@ function paypal_payment_pps_form_redirect_access(Payment $payment, $account = NU
...
-  PayPalPaymentIPNController::save($ipn);
-  $payment = entity_load_single('payment', $ipn->pid);
+function paypal_payment_pps_return(Payment $payment) {
+  if (PayPalPaymentIPNController::validate($_POST)) {
+    $status = PayPalPaymentIPNController::process($_POST);
+    $ipn = new PayPalPaymentIPN($_POST);
+    $ipn->pid = $status->pid;
+    $ipn->psiid = $status->psiid;
+    PayPalPaymentIPNController::save($ipn);
+  }
```
Comment (alex-mo) [inline paypal_payment_pps/paypal_payment_pps.module]: what happens if the `$_POST` does not validate?
maybe adding some `watchdog` log?

File: paypal_payment_pps/paypal_payment_pps.module [hunk 2]
```diff
@@ -247,26 +251,28 @@ function paypal_payment_pps_form_redirect_access(Payment $payment, $account = NU
...
+  // If the payment is still pending we assume we are returning from a
+  // successful payment.
+  if ($payment->getStatus()->status == PAYMENT_STATUS_PENDING) {
+    $payment->setStatus(new PaymentStatusItem(PAYMENT_STATUS_SUCCESS));
+    if (isset($_POST['txn_id'])) {
+      $ipn = new PayPalPaymentIPN($_POST);
```
Comment (alex-mo) [inline paypal_payment_pps/paypal_payment_pps.module]: missing validation for `$_POST` like in the `if` above?

File: paypal_payment_pps/paypal_payment_pps.module [hunk 3]
```diff
@@ -254,21 +254,23 @@ function paypal_payment_pps_form_redirect_access(Payment $payment, $account = NU
...
-      PayPalPaymentIPNController::save($ipn);
+    $status = new PaymentStatusItem(PAYMENT_STATUS_SUCCESS);
+    // After returning from PayPal the POST data might contain something that
+    // is not quite an IPN:
+    // - tx: The transaction id.
+    // - st: Status (usually "Pending").
+    // - amt: The total amount.
+    // - cc: The currency code.
+    // @see https://www.drupal.org/project/paypal_payment/issues/2918709#comment-12362825
+    if (isset($_POST['tx'])) {
```
Comment (alex-mo) [inline paypal_payment_pps/paypal_payment_pps.module:270]: "is not quite an IPN" does mean there might be a `txn_id` key sometimes?

**Review by alex-mo — CHANGES_REQUESTED** [review]
seems undefined what happens when the IPN payload does not validate...

File: paypal_payment_pps/paypal_payment_pps.module (hunk 3 above)
Comment (torotil) [reply paypal_payment_pps/paypal_payment_pps.module:270]: Is not quite an IPN means: What’s in the comment is what I have observed that paypal does. I haven’t found any docs or specs that tell me what to expect here.

**Review by alex-mo — APPROVED** [review]

**Pull request merged**